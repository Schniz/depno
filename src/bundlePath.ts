import { NodePath, Binding } from '@babel/traverse';
import { isChildScope } from './isChildScope';
import {
  program,
  Statement,
  expressionStatement,
  isStatement,
  Expression,
  Program,
  isImportNamespaceSpecifier,
  ImportDeclaration,
  isVariableDeclarator,
  variableDeclaration,
  isIdentifier,
  Identifier,
  isImportSpecifier,
  ImportSpecifier,
  File,
  isExportDefaultDeclaration,
  isFunctionDeclaration,
  isClassDeclaration,
  VariableDeclarator,
  FunctionDeclaration,
  variableDeclarator,
  identifier,
  objectPattern,
  objectProperty,
  callExpression,
  stringLiteral,
  CallExpression,
  templateLiteral,
  templateElement,
  isImportDefaultSpecifier,
  functionExpression,
  Node,
} from '@babel/types';
import * as types from '@babel/types';
import { getContentsFromURI } from './getContentsFromURI';
import { transformFromAstAsync, transformAsync } from '@babel/core';
import { resolve, dirname } from 'path';
import { resolve as urlResolve } from 'url';
import traverse from '@babel/traverse';
import generate from '@babel/generator';
import jsesc from 'jsesc';
import { globals } from './globals';

const nodeBuildinModules = ['fs', 'stream', 'http'];

function validateBinding(
  binding: Binding | undefined,
  path: NodePath<Identifier>,
  programPath: NodePath<Program>,
  pathToBundle: NodePath
) {
  if (!binding) {
    if (!globals.includes(path.node.name)) {
      // TODO: find a way to use buildCodeFrameError
      throw new ReferenceError(`Could not find ${path.node.name}`);
    } else {
      return false;
    }
  } else {
    if (binding.scope === programPath.scope) {
      if (isImportNamespaceSpecifier(binding.path.node)) {
        if (
          (binding.path.parent as ImportDeclaration).source.value === 'console'
        ) {
          return false;
        }
      }
      return true;
    } else {
      if (!isChildScope(binding.scope, pathToBundle.scope)) {
        throw path.buildCodeFrameError(
          `Cannot reference a non program declaration: ${path.node.name}`,
          ReferenceError
        );
      }
      return false;
    }
  }
}

export function getDefinitionNameFromNode(node: Node) {
  if (isVariableDeclarator(node)) {
    return (node.id as Identifier).name;
  } else if (types.isExportDefaultDeclaration(node)) {
    return 'default';
  } else if (isFunctionDeclaration(node)) {
    return node.id!.name;
  } else if (types.isVariableDeclaration(node)) {
    return (node.declarations[0].id as Identifier).name;
  } else if (isIdentifier(node)) {
    return node.name;
  }
  throw new Error(
    `Don't know how to getDefinitionNameFromNode for node of type ${node.type}`
  );
}

async function bundleDefinitionsForPath(
  pathToBundle: NodePath,
  programPath: NodePath<Program>,
  currentURI: string,
  wantedName?: string,
  bundledDefinitions: string[] = []
): Promise<Statement[]> {
  if (isIdentifier(pathToBundle.node)) {
    const binding = programPath.scope.getBinding(pathToBundle.node.name);
    if (
      validateBinding(
        binding,
        pathToBundle as NodePath<Identifier>,
        programPath,
        pathToBundle
      )
    ) {
      return bundleDefinitionsForPath(
        binding!.path,
        programPath,
        currentURI,
        undefined,
        bundledDefinitions
      );
    }
  }

  if (
    bundledDefinitions.includes(
      `${currentURI}#${getDefinitionNameFromNode(pathToBundle.node)}`
    )
  ) {
    return [];
  }

  bundledDefinitions.push(
    `${currentURI}#${getDefinitionNameFromNode(pathToBundle.node)}`
  );

  const outOfScopeBindings = new Set<Binding>();
  pathToBundle.traverse({
    // @ts-ignore
    ReferencedIdentifier(path: NodePath<Identifier>) {
      const binding = path.scope.getBinding(path.node.name);
      if (
        validateBinding(binding, path, programPath, pathToBundle) &&
        binding!.path !== pathToBundle
      ) {
        outOfScopeBindings.add(binding!);
      }
    },
  });

  const statements: Statement[] = [];

  for (const path of Array.from(outOfScopeBindings).map(
    binding => binding.path
  )) {
    if (isVariableDeclarator(path.node) || isFunctionDeclaration(path.node)) {
      statements.push(
        ...(await bundleDefinitionsForPath(
          path,
          programPath,
          currentURI,
          undefined,
          bundledDefinitions
        ))
      );
    } else if (
      isImportSpecifier(path.node) ||
      isImportDefaultSpecifier(path.node)
    ) {
      const dependencyPath = (path.parentPath as NodePath<ImportDeclaration>)
        .node.source.value;
      const dependencyURI = dependencyPath.startsWith('http://')
        ? dependencyPath
        : dependencyPath.startsWith('.')
        ? currentURI.startsWith('/')
          ? resolve(dirname(currentURI), dependencyPath)
          : urlResolve(currentURI, dependencyPath)
        : undefined;
      if (dependencyURI) {
        const code = await getContentsFromURI(dependencyURI);
        const { ast } = (await transformAsync(code, {
          filename: dependencyURI,
          ast: true,
          presets: [
            require('@babel/preset-typescript'),
            [
              require('@babel/preset-env'),
              {
                targets: {
                  node: 'current',
                },
                modules: false,
              },
            ],
          ],
        }))!;

        let dependencyNodePath: NodePath;
        let dependencyProgramPath: NodePath<Program>;
        traverse((ast as unknown) as File, {
          Program(programPath: NodePath<Program>) {
            dependencyProgramPath = programPath;
            let dependencyBinding;
            if (isImportDefaultSpecifier(path.node)) {
              programPath.traverse({
                ExportDefaultDeclaration: p => {
                  dependencyBinding = { path: p };
                },
              });
            } else {
              dependencyBinding = programPath.scope.getBinding(
                (path.node as ImportSpecifier).imported.name
              );
            }
            if (!dependencyBinding) {
              throw path.buildCodeFrameError(
                `Failed to find binding for ${
                  (path.node as ImportSpecifier).imported.name
                } at ${dependencyPath}`,
                ReferenceError
              );
            }
            dependencyNodePath = dependencyBinding.path;
          },
        });

        statements.push(
          ...(await bundleDefinitionsForPath(
            dependencyNodePath!,
            dependencyProgramPath!,
            dependencyURI,
            path.node.local.name,
            bundledDefinitions
          ))
        );
      } else if (nodeBuildinModules.includes(dependencyPath)) {
        statements.push(
          variableDeclaration('const', [
            variableDeclarator(
              objectPattern([
                objectProperty(path.node.local, path.node.local, false, true),
              ]),
              callExpression(identifier('require'), [
                stringLiteral(dependencyPath),
              ])
            ),
          ])
        );
      }
    } else if (isStatement(path.node)) {
      statements.push(path.node);
    } else {
      throw new Error('asdasdas');
    }
  }

  // when the path to bundle is a declartion in itself, it should be included
  // in the bindings
  if (isVariableDeclarator(pathToBundle.node)) {
    const binding = pathToBundle.scope.getBinding(
      ((pathToBundle.node as VariableDeclarator | FunctionDeclaration)
        .id as Identifier).name
    );
    if (
      validateBinding(
        binding,
        pathToBundle as NodePath<Identifier>,
        programPath,
        pathToBundle
      )
    ) {
      statements.push(variableDeclaration('const', [pathToBundle.node]));
    }
  } else if (isFunctionDeclaration(pathToBundle.node)) {
    statements.push(pathToBundle.node);
  } else if (isExportDefaultDeclaration(pathToBundle.node)) {
    if (wantedName) {
      statements.push(
        variableDeclaration('const', [
          variableDeclarator(
            identifier(wantedName),
            isFunctionDeclaration(pathToBundle.node.declaration)
              ? functionExpression(
                  pathToBundle.node.declaration.id,
                  pathToBundle.node.declaration.params,
                  pathToBundle.node.declaration.body,
                  pathToBundle.node.declaration.generator,
                  pathToBundle.node.declaration.async
                )
              : (pathToBundle.node.declaration as Expression)
          ),
        ])
      );
    }
  }

  return statements;
}

async function bundleIntoASingleProgram(
  pathToBundle: NodePath,
  programPath: NodePath<Program>,
  uri: string
) {
  const definitions = await bundleDefinitionsForPath(
    pathToBundle,
    programPath,
    uri
  );

  const pathValueStatement = isStatement(pathToBundle.node)
    ? isExportDefaultDeclaration(pathToBundle.node)
      ? isFunctionDeclaration(pathToBundle.node) ||
        isClassDeclaration(pathToBundle.node)
        ? pathToBundle.node
        : expressionStatement(pathToBundle.node.declaration as Expression)
      : pathToBundle.node
    : isVariableDeclarator(pathToBundle.node)
    ? expressionStatement(
        (pathToBundle.node as VariableDeclarator).id as Identifier
      )
    : expressionStatement(pathToBundle.node as Expression);

  return program(definitions.concat([pathValueStatement]));
}

export async function bundlePath(
  pathToBundle: NodePath,
  programPath: NodePath<Program>,
  currentURI: string
) {
  const programBeforeMacros = await bundleIntoASingleProgram(
    pathToBundle,
    programPath,
    currentURI
  );

  const { code } = (await transformFromAstAsync(
    programBeforeMacros!,
    undefined,
    {
      filename: currentURI,
      presets: [
        require('@babel/preset-typescript'),
        [
          require('@babel/preset-env'),
          {
            targets: ['current node'],
          },
        ],
      ],
      plugins: [
        () => ({
          visitor: {
            ReferencedIdentifier(path: NodePath<Identifier>, state: any) {
              if (path.node.name === 'bundleAST') {
                const callExpression = path.parentPath as NodePath<
                  CallExpression
                >;
                const toBundle = callExpression.get('arguments.0') as NodePath;
                const bundledProgram = bundleSameFile(
                  toBundle,
                  state.file.path
                );
                const code = generate(bundledProgram, {
                  compact: true,
                }).code;
                path.parentPath.replaceWith(
                  templateLiteral([templateElement({ raw: code })], [])
                );
              } else if (path.node.name === 'createMacro') {
                const macroVariableDeclaratorReferencePath =
                  path.parentPath.parentPath;
                const macroName = ((macroVariableDeclaratorReferencePath.node as VariableDeclarator)
                  .id as Identifier).name;
                const macroFunctionArgumentPath = path.parentPath.get(
                  'arguments.0'
                ) as NodePath<Expression>;

                const programForMacroArgument = bundleSameFile(
                  macroFunctionArgumentPath,
                  state.file.path
                );

                const code = generate(programForMacroArgument, {
                  compact: true,
                }).code;

                const macroFunction = eval(code);
                const macroDefinitionBinding = path.scope.getBinding(
                  macroName
                )!;
                for (const macroReference of macroDefinitionBinding.referencePaths) {
                  macroFunction({ reference: macroReference, types, state });
                  macroVariableDeclaratorReferencePath.remove();
                }
              }
            },
          },
        }),
      ],
    }
  ))!;

  return jsesc(code!, { quotes: 'backtick' });
}

function bundleSameFile(
  pathToBundle: NodePath,
  programPath: NodePath<Program>
) {
  const outOfScopeBindings = new Set<Binding>();
  if (isIdentifier(pathToBundle.node)) {
    const binding = programPath.scope.getBinding(pathToBundle.node.name);
    outOfScopeBindings.add(binding!);
  } else {
    // when the path to bundle is a declartion in itself, it should be included
    // in the bindings
    if (isVariableDeclarator(pathToBundle.node)) {
      const binding = pathToBundle.scope.getBinding(
        ((pathToBundle.node as VariableDeclarator).id as Identifier).name
      );
      outOfScopeBindings.add(binding!);
    } else if (isFunctionDeclaration(pathToBundle.node)) {
      const binding = pathToBundle.scope.getBinding(
        (pathToBundle.node as FunctionDeclaration).id?.name
      );
      outOfScopeBindings.add(binding!);
    }

    pathToBundle.traverse({
      // @ts-ignore
      ReferencedIdentifier(path: NodePath<Identifier>) {
        const binding = path.scope.getBinding(path.node.name);
        if (validateBinding(binding, path, programPath, pathToBundle)) {
          outOfScopeBindings.add(binding!);
        }
      },
    });
  }

  const statements: Statement[] = [];

  for (const path of Array.from(outOfScopeBindings).map(
    binding => binding.path
  )) {
    if (isVariableDeclarator(path.node)) {
      statements.push(variableDeclaration('const', [path.node]));
    } else if (isStatement(path.node)) {
      statements.push(path.node);
    } else {
      throw new Error('asdasdas');
    }
  }

  const pathValueStatement = isStatement(pathToBundle.node)
    ? isExportDefaultDeclaration(pathToBundle.node)
      ? isFunctionDeclaration(pathToBundle.node) ||
        isClassDeclaration(pathToBundle.node)
        ? pathToBundle.node
        : expressionStatement(pathToBundle.node.declaration as Expression)
      : pathToBundle.node
    : isVariableDeclarator(pathToBundle.node)
    ? expressionStatement(
        (pathToBundle.node as VariableDeclarator).id as Identifier
      )
    : expressionStatement(pathToBundle.node as Expression);

  return program(statements.concat([pathValueStatement]));
}
