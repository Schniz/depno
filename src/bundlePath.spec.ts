import traverse, { NodePath } from '@babel/traverse';
import {
  VariableDeclarator,
  Program,
  Identifier,
  isStatement,
} from '@babel/types';
import { bundlePath } from './bundlePath';
import { parseSync } from '@babel/core';
import { tuple } from '@deaven/tuple';

describe(bundlePath, () => {
  test('should bundle a file scoped binding', async () => {
    const code = `
      function a() {

      }
      const b = 6;
      const c = a;
		`;

    const [
      pathToBundle,
      programPath,
    ] = extractPathToBundleAndProgramPathFromCode(code, 'c');

    expect(
      await bundlePath(pathToBundle!, programPath!, '/a.ts')
    ).toMatchInlineSnapshot(`"function a() {}\\\\n\\\\nconst c = a;"`);
  });

  test('should bundle variable declarations', async () => {
    const code = `
      const b = 1;
      const c = b;
		`;

    const [
      pathToBundle,
      programPath,
    ] = extractPathToBundleAndProgramPathFromCode(code, 'c');

    expect(
      await bundlePath(pathToBundle!, programPath!, '/a.ts')
    ).toMatchInlineSnapshot(`"const b = 1;\\\\nconst c = b;"`);
  });

  test('should bundle transitive variable declarations', async () => {
    const code = `
		  const a = 1;
      const b = a;
      const c = b;
		`;

    const [
      pathToBundle,
      programPath,
    ] = extractPathToBundleAndProgramPathFromCode(code, 'c');

    expect(
      await bundlePath(pathToBundle!, programPath!, '/a.ts')
    ).toMatchInlineSnapshot(`"const a = 1;\\\\nconst b = a;\\\\nconst c = b;"`);
  });

  test('should bundle identifier', async () => {
    const code = `
      const b = 1;
      b
		`;

    const ast = parseSync(code, {
      filename: 'a.ts',
    })!;

    let pathToBundle: NodePath;
    let programPath: NodePath<Program>;
    traverse(ast, {
      Program: path => {
        programPath = path;
      },
      Identifier: function(path) {
        if (isStatement(path.parentPath.node)) {
          pathToBundle = path;
        }
      },
    });

    expect(
      await bundlePath(pathToBundle!, programPath!, '/a.ts')
    ).toMatchInlineSnapshot(`"const b = 1;\\\\nb;"`);
  });

  test('should throw an error on undeclared reference', async () => {
    const code = `
      const b = 6;
      const c = a;
    `;

    const [
      pathToBundle,
      programPath,
    ] = extractPathToBundleAndProgramPathFromCode(code, 'c');

    await expect(
      bundlePath(pathToBundle!, programPath!, '/a.ts')
    ).rejects.toThrowError();
  });

  test('should throw an error on a reference to a non program scoped binding', async () => {
    const code = `
      const b = 6;
      const c = a;

      function d() {
        const e = 6;
        function f() {
          const h = e;
        }
      }
		`;

    const [
      pathToBundle,
      programPath,
    ] = extractPathToBundleAndProgramPathFromCode(code, 'h');

    await expect(
      bundlePath(pathToBundle!, programPath!, '/a.ts')
    ).rejects.toThrowError();
  });

  describe('builtin modules', () => {
    test('importing the console module', async () => {
      const code = `
				import * as console from "console";

				const b = console;
			`;

      const [
        pathToBundle,
        programPath,
      ] = extractPathToBundleAndProgramPathFromCode(code, 'b');

      expect(
        await bundlePath(pathToBundle!, programPath!, '/a.ts')
      ).toMatchInlineSnapshot(`"const b = console;"`);
    });
  });
});

function extractPathToBundleAndProgramPathFromCode(
  code: string,
  identifier: string
) {
  const ast = parseSync(code, {
    filename: 'a.ts',
  })!;

  let pathToBundle: NodePath;
  let programPath: NodePath<Program>;
  traverse(ast, {
    Program: path => {
      programPath = path;
    },
    VariableDeclaration: function(path) {
      if (
        ((path.get('declarations.0') as NodePath<VariableDeclarator>).node
          .id as Identifier).name === identifier
      ) {
        pathToBundle = path;
      }
    },
  });

  return tuple(pathToBundle, programPath);
}
