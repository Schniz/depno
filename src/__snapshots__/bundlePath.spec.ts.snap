// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bundlePath builtin modules importing the console module 1`] = `
Object {
  "body": Array [
    Node {
      "declarations": Array [
        Node {
          "end": 63,
          "id": Node {
            "end": 53,
            "loc": SourceLocation {
              "end": Position {
                "column": 11,
                "line": 4,
              },
              "identifierName": "b",
              "start": Position {
                "column": 10,
                "line": 4,
              },
            },
            "name": "b",
            "start": 52,
            "type": "Identifier",
          },
          "init": Node {
            "end": 63,
            "loc": SourceLocation {
              "end": Position {
                "column": 21,
                "line": 4,
              },
              "identifierName": "console",
              "start": Position {
                "column": 14,
                "line": 4,
              },
            },
            "name": "console",
            "start": 56,
            "type": "Identifier",
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 21,
              "line": 4,
            },
            "start": Position {
              "column": 10,
              "line": 4,
            },
          },
          "start": 52,
          "type": "VariableDeclarator",
        },
      ],
      "end": 64,
      "kind": "const",
      "loc": SourceLocation {
        "end": Position {
          "column": 22,
          "line": 4,
        },
        "start": Position {
          "column": 4,
          "line": 4,
        },
      },
      "start": 46,
      "type": "VariableDeclaration",
    },
  ],
  "directives": Array [],
  "interpreter": null,
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`bundlePath should bundle a file scoped binding 1`] = `
Object {
  "body": Array [
    Node {
      "async": false,
      "body": Node {
        "body": Array [],
        "directives": Array [],
        "end": 30,
        "loc": SourceLocation {
          "end": Position {
            "column": 7,
            "line": 4,
          },
          "start": Position {
            "column": 19,
            "line": 2,
          },
        },
        "start": 20,
        "type": "BlockStatement",
      },
      "end": 30,
      "generator": false,
      "id": Node {
        "end": 17,
        "loc": SourceLocation {
          "end": Position {
            "column": 16,
            "line": 2,
          },
          "identifierName": "a",
          "start": Position {
            "column": 15,
            "line": 2,
          },
        },
        "name": "a",
        "start": 16,
        "type": "Identifier",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 7,
          "line": 4,
        },
        "start": Position {
          "column": 6,
          "line": 2,
        },
      },
      "params": Array [],
      "start": 7,
      "type": "FunctionDeclaration",
    },
    Node {
      "declarations": Array [
        Node {
          "end": 67,
          "id": Node {
            "end": 63,
            "loc": SourceLocation {
              "end": Position {
                "column": 13,
                "line": 6,
              },
              "identifierName": "c",
              "start": Position {
                "column": 12,
                "line": 6,
              },
            },
            "name": "c",
            "start": 62,
            "type": "Identifier",
          },
          "init": Node {
            "end": 67,
            "loc": SourceLocation {
              "end": Position {
                "column": 17,
                "line": 6,
              },
              "identifierName": "a",
              "start": Position {
                "column": 16,
                "line": 6,
              },
            },
            "name": "a",
            "start": 66,
            "type": "Identifier",
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 17,
              "line": 6,
            },
            "start": Position {
              "column": 12,
              "line": 6,
            },
          },
          "start": 62,
          "type": "VariableDeclarator",
        },
      ],
      "end": 68,
      "kind": "const",
      "loc": SourceLocation {
        "end": Position {
          "column": 18,
          "line": 6,
        },
        "start": Position {
          "column": 6,
          "line": 6,
        },
      },
      "start": 56,
      "type": "VariableDeclaration",
    },
  ],
  "directives": Array [],
  "interpreter": null,
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`bundlePath should bunlde variable declarations 1`] = `
Object {
  "body": Array [
    Object {
      "declarations": Array [
        Node {
          "end": 18,
          "id": Node {
            "end": 14,
            "loc": SourceLocation {
              "end": Position {
                "column": 13,
                "line": 2,
              },
              "identifierName": "b",
              "start": Position {
                "column": 12,
                "line": 2,
              },
            },
            "name": "b",
            "start": 13,
            "type": "Identifier",
          },
          "init": Node {
            "end": 18,
            "extra": Object {
              "raw": "1",
              "rawValue": 1,
            },
            "loc": SourceLocation {
              "end": Position {
                "column": 17,
                "line": 2,
              },
              "start": Position {
                "column": 16,
                "line": 2,
              },
            },
            "start": 17,
            "type": "NumericLiteral",
            "value": 1,
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 17,
              "line": 2,
            },
            "start": Position {
              "column": 12,
              "line": 2,
            },
          },
          "start": 13,
          "type": "VariableDeclarator",
        },
      ],
      "kind": "const",
      "type": "VariableDeclaration",
    },
    Node {
      "declarations": Array [
        Node {
          "end": 37,
          "id": Node {
            "end": 33,
            "loc": SourceLocation {
              "end": Position {
                "column": 13,
                "line": 3,
              },
              "identifierName": "c",
              "start": Position {
                "column": 12,
                "line": 3,
              },
            },
            "name": "c",
            "start": 32,
            "type": "Identifier",
          },
          "init": Node {
            "end": 37,
            "loc": SourceLocation {
              "end": Position {
                "column": 17,
                "line": 3,
              },
              "identifierName": "b",
              "start": Position {
                "column": 16,
                "line": 3,
              },
            },
            "name": "b",
            "start": 36,
            "type": "Identifier",
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 17,
              "line": 3,
            },
            "start": Position {
              "column": 12,
              "line": 3,
            },
          },
          "start": 32,
          "type": "VariableDeclarator",
        },
      ],
      "end": 38,
      "kind": "const",
      "loc": SourceLocation {
        "end": Position {
          "column": 18,
          "line": 3,
        },
        "start": Position {
          "column": 6,
          "line": 3,
        },
      },
      "start": 26,
      "type": "VariableDeclaration",
    },
  ],
  "directives": Array [],
  "interpreter": null,
  "sourceType": "script",
  "type": "Program",
}
`;
